### PR TITLE
Planning: use path point guide for speed limit

### PR DESCRIPTION
**Pull request by JasonZhou404**

### CODE REVIEW

File: modules/planning/tasks/optimizers/dp_st_speed/dp_st_cost.cc [hunk 1]
```diff
@@ -189,14 +189,17 @@ double DpStCost::GetSpeedCost(const STPoint& first, const STPoint& second,
...
-    cost += config_.exceed_soft_speed_penalty() * config_.default_speed_cost() *
-            fabs(speed * speed) * unit_t_;
-  } else if (soft_det_speed < 0) {
-    cost += config_.low_soft_speed_penalty() * config_.default_speed_cost() *
-            -soft_det_speed * unit_t_;
+  if (FLAGS_enable_soft_speed_limit) {
+    double soft_det_speed = (speed - soft_speed_limit) / soft_speed_limit;
+    if (soft_det_speed > 0) {
+      cost += config_.exceed_soft_speed_penalty() *
+              config_.default_speed_cost() * fabs(speed * speed) * unit_t_;
```
Comment (YajiaZhang) [inline modules/planning/tasks/optimizers/dp_st_speed/dp_st_cost.cc]: Should here use "speed" or "speed - soft_speed_limit" when you are trying to penalize the exceeded speed?

Comment (JasonZhou404) [reply modules/planning/tasks/optimizers/dp_st_speed/dp_st_cost.cc]: I just copied how speed cost calculated before. Will improve it if it's needed when I debug and test it in next step

File: modules/planning/tasks/deciders/speed_bounds_decider/speed_limit_decider.cc [hunk 2]
```diff
@@ -138,47 +149,77 @@ Status SpeedLimitDecider::GetSpeedLimits(
...
+                const_obstacle->PerceptionSLBoundary().end_s()) {
+          continue;
+        }
+      } else {
+        if (frenet_point_s + vehicle_param_.front_edge_to_center() <
+                const_obstacle->PerceptionSLBoundary().start_s() ||
+            frenet_point_s - vehicle_param_.back_edge_to_center() >
+                const_obstacle->PerceptionSLBoundary().end_s()) {
+          continue;
+        }
```
Comment (xiaoxq) [inline modules/planning/tasks/deciders/speed_bounds_decider/speed_limit_decider.cc]: DRY: Don't repeat yourself.

File: modules/planning/tasks/optimizers/dp_st_speed/dp_st_graph.cc [hunk 3]
```diff
@@ -259,8 +259,11 @@ void DpStGraph::CalculateCostAt(const std::shared_ptr<StGraphMessage>& msg) {
 
   double speed_limit =
       st_graph_data_.speed_limit().GetSpeedLimitByS(unit_s_ * r);
-  double soft_speed_limit =
-      st_graph_data_.speed_limit().GetSoftSpeedLimitByS(unit_s_ * r);
+  double soft_speed_limit = speed_limit;
+  if (FLAGS_enable_soft_speed_limit) {
+    soft_speed_limit =
+        st_graph_data_.speed_limit().GetSoftSpeedLimitByS(unit_s_ * r);
+  }
```
Comment (xiaoxq) [inline modules/planning/tasks/optimizers/dp_st_speed/dp_st_graph.cc]: const double soft_speed_limit = FLAGS_enable_soft_speed_limit ? A : B;

**Pull request merged**